v03_pipeline/lib/misc/allele_registry.py

```diff
@@ -0,0 +1,169 @@
+import dataclasses
+import hashlib
+import math
+import tempfile
+import time
+
+import hail as hl
+import requests
+from requests import HTTPError
+
+from v03_pipeline.lib.logger import get_logger
+from v03_pipeline.lib.model import Env, ReferenceGenome
+
+MAX_VARIANTS_PER_REQUEST = 1000000
+ALLELE_REGISTRY_URL = 'https://reg.genome.network/alleles?file=vcf&fields=none+@id+externalRecords.gnomAD_4.id'
+HTTP_REQUEST_TIMEOUT = 300
+
+logger = get_logger(__name__)
+
+
+@dataclasses.dataclass
+class AlleleRegistryError:
+    base_url: str
+    error_type: str
+    description: str
+    message: str
+    input_line: str | None
+
+    @classmethod
+    def from_api_response(cls, response: dict, base_url: str):
+        return cls(
+            base_url=base_url,
+            error_type=response['errorType'],
+            description=response['description'],
+            message=response['message'],
+            input_line=response.get('inputLine'),
+        )
+
+    def __str__(self) -> str:
+        msg = (
+            f'\nAPI URL: {self.base_url}\nTYPE: {self.error_type}'
+            f'\nDESCRIPTION: {self.description}\nMESSAGE: {self.message}'
+        )
+        return (
+            msg if self.input_line is None else f'{msg}\nINPUT_LINE: {self.input_line}'
+        )
+
+
+def register_alleles_in_chunks(
+    ht: hl.Table,
+    reference_genome: ReferenceGenome,
+    base_url: str = ALLELE_REGISTRY_URL,
+    chunk_size: int = MAX_VARIANTS_PER_REQUEST,
+):
+    num_rows = ht.count()
+    num_chunks = math.ceil(num_rows / chunk_size)
+    logger.info(
+        f'Registering {num_rows} allele(s) in chunks of {chunk_size} in {num_chunks} request(s).',
+    )
+    for start_idx in range(0, num_rows, chunk_size):
+        end_idx = start_idx + chunk_size
+        if end_idx == chunk_size:
+            chunk_ht = ht.head(chunk_size)
```

Comment:
this code is super cool and would be a great addition to hail.

Reply:
I got the idea from this [discussion](https://discuss.hail.is/t/way-for-slicing-on-hail-table/2841/2) - it seems like it's kind of a hail anti-pattern (and an expensive operation?) to chunk out tables like this. 

(I tried the suggestion from the discussion to split the table by chromosomes using `parse_locus_interval` but there's way more than 1 million variants per chromosome, so it didn't seem like the right solution for our case.)
```
filtered = hl.filter_intervals(s37_ht, [hl.parse_locus_interval('1-2', reference_genome='GRCh37')])
# 7 million variants in chrom 1, this solution won't work
```

Reply:
oh yes it's definitely a hail anti pattern to localize 🤣 

could you do something like `repartition` to many small pieces and then `export` to tsv with the `header_per_shard` option to get exported chunks?  I guess that's not too dissimilar from exporting to vcf.

Reply:
This is a good idea and probably the sparkier way to do it but I'd rather keep the solution as it is now instead of refactoring both the chunking and exporting (tsv vs vcf) parts of the code